### DISCUSSION TITLE
Unable to display JSON record retrieved from SWR

### DISCUSSION THREAD

**Discussion by kendjr** [vercel/next.js#40824] [Help] [answered]
I am receiving an error when trying to display part of a JSON record retrieved using SWR, however, I am able to display the record in JSON format.

> TypeError: Cannot read properties of undefined (reading 'name')

The file testswr.js compiles and displays the entire JSON record:

```
import useSWR from "swr";
const fetcher = (url) => fetch(url).then((res) => res.json());

export default function Index() {
  const { data, error } = useSWR("/api/staticdata", fetcher);

  if (error) return <div>Failed to load</div>;
  if (!data) return <div>Loading...</div>;
  console.log("data is: " + data);
  return (
    <div>
      <h1>My Framework from file</h1>
      <ul>
        <li>Name: {data.record}</li>
        <li>testrecordname: {data}</li>
      </ul>
    </div>
  );
}

```
The code above displays:

```
My Framework from file
Name:
testrecordname: { "record": { "id": 8221, "uid": "a15c1f1d-9e4e-4dc7-9c45-c04412fc5064", "name": "Next.js", "language": "JavaScript" } }
```

This is a next.js application. I am using the code verbatim from https://vercel.com/guides/loading-static-file-nextjs-api-route.

Any help would be greatly appreciated!

**Comment by icyJoseph** [comment, marked as answer]
What do you get if you do `console.log(typeof data)`, I suspect it is a string, which is indeed valid JSON, we are often most familiar with its object shape, or array shape, but strings and numbers are perfectly valid JSON. In your console you can do `JSON.parse("2")` and it'll return Number 2 😅 

One additional thing is that, to debug this sort of thing, it is best to remove as many moving parts as possible. That means that the React + SWR layer is getting on the way here.

From your Browser Console or any REST Client, you can make a request for `fetch("/api/staticdata").then((res) => res.json())`, and try to debug it from there.

For example, from my console I've done:

```js
fetch("/api/staticdata").then(res=>res.json()).then(res => console.log(res, typeof res))
```

And it logs:

```
Promise {<pending>}
  "record": {
    "id": 8221,
    "uid": "a15c1f1d-9e4e-4dc7-9c45-c04412fc5064",
    "name": "Next.js",
    "language": "JavaScript"
  }
}
 string
```

So, bingo, it is a string! What gives?

Let's think a bit here:

```js
export default async function handler(req, res) {
  //Find the absolute path of the json directory
  const jsonDirectory = path.join(process.cwd(), "json");
  //Read the json data file data.json
  const fileContents = await fs.readFile(jsonDirectory + "/data.json", "utf8");
  // 🤔🤔🤔🤔🤔
  // At this point, fileContents is a string, and as said before, 
  // strings are somewhat valid JSON - so it is send as is
  res.status(200).json(fileContents);
}
```

Let's tweak this a bit:

```js
export default async function handler(req, res) {
  //Find the absolute path of the json directory
  const jsonDirectory = path.join(process.cwd(), "json");
  //Read the json data file data.json
  const fileContents = await fs.readFile(jsonDirectory + "/data.json", "utf8");
  const obj = JSON.parse(fileContents);
  //Return the content of the data file in json format
  res.status(200).json(obj);
}
```

And now we query from the browser console:

```js
fetch("/api/staticdata").then(res=>res.json()).then(res => console.log(res, typeof res))
```

And it logs:

```shell
Promise {<pending>}
{record: {…}} 'object'
```

> Note that you could `JSON.parse` the result on the client as well, but why block the UI parsing JSON if it can be done server side?

**Reply by kendjr** [reply]
Thanks!  Extremely educational explanation.  Data was indeed a string, and after I used JSON.parse I was able to correctly use it as a JSON object.

Would you recommend SWR for this type of data retrieval, or is it better to simply use fetch as you have above?

Thanks!
Ken

**Reply by icyJoseph** [reply]
Yeah, I think SWR is alright here, the new (beta) React Docs go on great details about `useEffect+useState` and data-fetching libraries: 

- https://beta.reactjs.org/learn/you-might-not-need-an-effect
- https://beta.reactjs.org/learn/synchronizing-with-effects#fetching-data

**Reply by lvanoverberghe** [reply]
I'm running into the same problem using the same example linked on https://vercel.com/guides/loading-static-file-nextjs-api-route. I changed the code of the API to parse the string, but now the loading variable returned by useSwr keeps on loading state.
